### PR TITLE
[libcxx] Allow string to use SSO in constant evaluation.

### PR DESCRIPTION
**Pull request by jyknight**
Previously, libcxx forced all strings created during constant evaluation to point to allocated memory. That was done due to implementation difficultites, but it turns out not to be necessary. This patch permits the use of SSO strings during constant evaluation, and also simplifies the implementation.

This does have a downside in terms of enabling users to accidentally write non-portable code, however, which I've documented in UsingLibcxx.rst.

In particular, whether `constinit std::string x = "...";` will successfully compile now depends on whether the string is smaller than the SSO capacity -- in libc++, up to 22 bytes on 64-bit platforms, and up to 10 bytes on 32-bit platforms. By comparison, libstdc++ and MSVC have an SSO capacity of 15 bytes, except that in libstdc++, constant-initialized strings cannot be used as function-locals because the object contains a pointer to itself.

### CODE REVIEW

**Comment by MaskRay** [comment]
Is there a test to demonstrate the new behavior?

File: libcxx/include/string [hunk 1]
```diff
@@ -1748,24 +1750,11 @@ private:
...
-        }
-    }
-
-    _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 void __deallocate_constexpr() {
-        if (__libcpp_is_constant_evaluated() && __get_pointer() != nullptr)
-            __alloc_traits::deallocate(__alloc(), __get_pointer(), __get_long_cap());
     }
 
     _LIBCPP_CONSTEXPR _LIBCPP_HIDE_FROM_ABI static bool __fits_in_sso(size_type __sz) {
         // SSO is disabled during constant evaluation because `__is_long` isn't constexpr friendly
```
Comment (frederick-vs-ja) [inline libcxx/include/string:1756]:
```suggestion
```
The comment in line should be removed.

File: libcxx/include/string [hunk 2]
```diff
@@ -1729,8 +1729,10 @@ private:
 
     _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20
     bool __is_long() const _NOEXCEPT {
-        if (__libcpp_is_constant_evaluated())
-            return true;
+        if (__libcpp_is_constant_evaluated()) {
+            if (__builtin_constant_p(__r_.first().__l.__is_long_))
+                return __r_.first().__l.__is_long_;
+        }
```
Comment (philnik777) [inline libcxx/include/string:1735]:
```suggestion
        if (__libcpp_is_constant_evaluated() && __builtin_constant_p(__r_.first().__l.__is_long_)) {
            return __r_.first().__l.__is_long_;
        }
```

File: libcxx/include/string [hunk 3]
```diff
@@ -1748,24 +1750,11 @@ private:
 
     _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 void __default_init() {
         __r_.first() = __rep();
```
Comment (philnik777) [inline libcxx/include/string:1752]: I think we can just inline this now. No need to keep the function for this.

File: libcxx/include/string [hunk 4]
```diff
@@ -1969,7 +1955,8 @@ private:
                     allocator_type __a = __str.__alloc();
                     auto __allocation = std::__allocate_at_least(__a, __str.__get_long_cap());
                     __begin_lifetime(__allocation.ptr, __allocation.count);
-                    __alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap());
+                    if (__is_long())
+                        __alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap());
```
Comment (philnik777) [inline libcxx/include/string:1959]: Why is this change required? AFAICT there should be no changes that affect the runtime.

**Review by philnik777 — CHANGES_REQUESTED** [review]
In general I really like this patch. I understand that this could be a portability problem, but given that libstdc++ and the MSVC STL already allow this, I don't think that's something we introduce. Maybe we could add a warning for this case. I think that would address any portability concerns. @AaronBallman @cor3ntin @erichkeane do you have any idea if that would be feasible?

File: libcxx/include/string (hunk 4 above)
Comment (jyknight) [reply libcxx/include/string:1959]: Looks like this is a pre-existing bug. I changed this because it failed the test-suite without this change, as it was trying to call get_long_pointer on a short string during constant evaluation. I didn't look closely enough before to see that it's actually reachable in normal code too, if you have an allocator with `propagate_on_container_copy_assignment`, and which has unequal instances, and you are assigning from a long string to a short string that previously contained more than 15 characters (such that the long-pointer didn't happen to be zero).

Demo of crash: https://godbolt.org/z/TqEf15K3K

File: libcxx/include/string (hunk 3 above)
Comment (jyknight) [reply libcxx/include/string:1752]: Done.

File: libcxx/include/string (hunk 2 above)
Comment (jyknight) [reply libcxx/include/string:1735]: Done.

File: libcxx/include/string [hunk 5]
```diff
@@ -830,8 +830,8 @@ private:
     {
         union
         {
-            __long  __l;
             __short __s;
+            __long  __l;
             __raw   __r;
```
[the remark this answers is not in the document]
Comment (jyknight) [reply libcxx/include/string:835]: The use of `if (__libcpp_is_constant_evaluated()) __r_.first() = __rep();` throughout the code will activate the first alternative in the union. Now, we want that default-active alternative to be the short repr, so I moved it to be first.

File: libcxx/include/string (hunk 1 above)
Comment (jyknight) [reply libcxx/include/string:1756]: Done.

**Comment by jyknight** [comment]
> Is there a test to demonstrate the new behavior?

I have not written one at the moment, no.

File: libcxx/include/string (hunk 4 above)
Comment (philnik777) [reply libcxx/include/string:1953]: That looks like a pretty serious bug to me. Could you move this change into it's own PR with the test-case added?

File: libcxx/include/string [hunk 6]
```diff
@@ -880,7 +880,7 @@ public:
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 basic_string()
       _NOEXCEPT_(is_nothrow_default_constructible<allocator_type>::value)
       : __r_(__default_init_tag(), __default_init_tag()) {
-    __default_init();
+    __r_.first() = __rep();
```
Comment (philnik777) [inline libcxx/include/string:883]: I think this could just be
```suggestion
  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 basic_string()
      _NOEXCEPT_(is_nothrow_default_constructible<allocator_type>::value)
      : __r_(__value_init_tag(), __default_init_tag()) {}
```
Same for the change below.

File: libcxx/include/string [hunk 7]
```diff
@@ -1746,26 +1747,8 @@ private:
...
-
-    _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 void __deallocate_constexpr() {
-        if (__libcpp_is_constant_evaluated() && __get_pointer() != nullptr)
-            __alloc_traits::deallocate(__alloc(), __get_pointer(), __get_long_cap());
-    }
-
     _LIBCPP_CONSTEXPR _LIBCPP_HIDE_FROM_ABI static bool __fits_in_sso(size_type __sz) {
-        // SSO is disabled during constant evaluation because `__is_long` isn't constexpr friendly
-        return !__libcpp_is_constant_evaluated() && (__sz < __min_cap);
+        return (__sz < __min_cap);
```
Comment (philnik777) [inline libcxx/include/string:1751]:
```suggestion
        return __sz < __min_cap;
```

**Review by philnik777 — CHANGES_REQUESTED** [review]
I think this looks quite good. I'd like a libc++-specific test for something simple like `constinit std::string my_str = "";` and a short paragraph in `libcxx/docs/UsingLibcxx.rst` about this. Maybe something at the bottom about unspecified behaviour which we define.

**Comment by erichkeane** [comment]
> In general I really like this patch. I understand that this could be a portability problem, but given that libstdc++ and the MSVC STL already allow this, I don't think that's something we introduce. Maybe we could add a warning for this case. I think that would address any portability concerns. @AaronBallman @cor3ntin @erichkeane do you have any idea if that would be feasible?

I'm not quite sure the question here, `#warning` exists, but otherwise I'm not sure what type of warning you're looking for.

**Comment by jyknight** [comment]
> Maybe we could add a warning for this case. I think that would address any portability concerns.

That's a great idea!

> I'm not quite sure the question here, `#warning` exists, but otherwise I'm not sure what type of warning you're looking for.

We want a warning if an object of a particular type (std::basic_string) is initialized at constant-evaluation-time, and then the resulting object escapes and is embedded into the resulting binary for use at runtime. The warning is justified for std::basic_string, because escaping the object to runtime is only possible (at least so far), when it doesn't point to any external memory allocations (and also it must not point to itself, when it's a local constinit variable) -- which are all internal implementation details. Whether a string meets these constraints is thus implementation-defined, and in practice varies based on platform and standard-lib implementation, and of course length of the string.

We could make a generic type attribute to diagnose this, or perhaps just hard-code it for std::basic_string.

**Comment by cor3ntin** [comment]
This looks like a good motivation for both [Exposing the diagnostic engine to C++](https://discourse.llvm.org/t/exposing-the-diagnostic-engine-to-c/73092) (LLVM) and [P2758 Emitting messages at compile time](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2758r0.html). 

And of course, folks are still working on allocation propagation which would make the need for a warning here moot.

I'm not sure how easy it would be for us to track the uses of basic strings in context that are potentially evaluated and not constant evaluated. For example what about constexpr functions that themselves are only used at compile time?

**Comment by philnik777** [comment]
> This looks like a good motivation for both [Exposing the diagnostic engine to C++](https://discourse.llvm.org/t/exposing-the-diagnostic-engine-to-c/73092) (LLVM) and [P2758 Emitting messages at compile time](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2758r0.html).

I'm not sure how that would help. The problem here is that we have to detect whether the string is promoted to static storage, which AFAIK isn't possible to detect in this case.

**Comment by frederick-vs-ja** [comment]
I think the warning should be triggered only if
- a `constexpr` or `constinit` object is a `basic_string` or contains a `basic_string` subobject, or
- the definition of a `constexpr` or `constinit` variable extends the lifetime of a temporary object that meets the previous condition.

For a `basic_string` variable of static/thread storage duration that is not declared with `constexpr` or `constinit` and happens to be constant-initialized, the warning seems undesired - as the definition itself is portable.

File: libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
+// Ensure that strings which fit within the SSO size can be constant-initialized
```
Comment (shafik) [inline libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp:11]: I think I would also like to see some Sema tests as well, do we expect something like this to work: https://godbolt.org/z/x8EvdGfoq

Currently it does not for libstdc++ but does on MSVC.

```cpp
#include <string>

constexpr std::string f() {
    std::string s = "hello world";

    return s;
}

void g() {
    constexpr auto s = f();
}
```

**Review by shafik — COMMENTED** [review]
I think this is fine but I agree we should have a compat warning for this and I would like to see more testing as well.

**Review by philnik777 — APPROVED** [review]
LGTM % nit.

**Comment by ldionne** [comment]
I just want to slide in to say that I *really* like this patch from the point of view of `std::string`'s implementation. Our `__default_init()` method and the `if (is_constant_evaluated())` were pretty awful, and I think it's great to remove them.

I agree it could cause a portability trap to enable this, but I think the right way to go is to have a nice Clang diagnostic that says "oops, strings are can't portably be initialized at compile-time" -- not to make our `std::string` implementation go out of its way to be pessimized at compile-time.

Thanks for working on that @jyknight!

**Comment by ldionne** [comment]
@jyknight Is there anything you need in order to make progress on this? If so, please let us know and we'll try to help. I think this patch is pretty much ready to go based on @philnik777 's comment above.

**Comment by jyknight** [comment]
I don't think I will be able to work on that the discussed Clang diagnostics anytime soon. Unless someone else is volunteering to implement it, that may not be implemented at all.

So, I'd like to hear explicit consensus is that folks are generally OK to submit this _without_ such a diagnostic.

**Comment by ldionne** [comment]
> I don't think I will be able to work on that the discussed Clang diagnostics anytime soon. Unless someone else is volunteering to implement it, that may not be implemented at all.
> 
> So, I'd like to hear explicit consensus is that folks are generally OK to submit this _without_ such a diagnostic.

Oh, I might have misread or misunderstood the discussion, but I didn't have in mind that we would block this patch from making progress on having a Clang diagnostic. Did other people in this conversation have that expectation?

**Comment by philnik777** [comment]
I didn't consider that blocking either. IMO that's actually out-of-scope for this patch.

**Comment by AaronBallman** [comment]
> > I don't think I will be able to work on that the discussed Clang diagnostics anytime soon. Unless someone else is volunteering to implement it, that may not be implemented at all.
> > So, I'd like to hear explicit consensus is that folks are generally OK to submit this _without_ such a diagnostic.
> 
> Oh, I might have misread or misunderstood the discussion, but I didn't have in mind that we would block this patch from making progress on having a Clang diagnostic. Did other people in this conversation have that expectation?
> 
> Edit: In other words, I think we should ship this and then adding a warning is a nice-to-have followup to reduce the likelihood of people falling into a portability trap.

I think that's a reasonable path forward.

File: libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp (hunk 8 above)
Comment (jyknight) [reply libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp:11]: We do expect that to work. It doesn't work on GCC because libstdc++'s std::string has a self-pointer, and a self-pointer is only constant for a global (or function-local static), not a local variable.

Added a test that covers the local var case.

File: libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+// globals.  (this is permitted but not required to work by the standard).
```
[the remark this answers is not in the document]
Comment (jyknight) [reply libcxx/test/libcxx/strings/basic.string/string.cons/constinit_sso_string.compile.pass.cpp]: Done.

**Comment by jyknight** [comment]
Looks like the debug/hardened/safe mode tests are failing libcxx/std/utilities/template.bitset/bitset.members/to_string.pass.cpp because the constant evaluations for
```
  static_assert(test_to_string<64>());
  static_assert(test_to_string<65>());
```
now both hit the constant evaluation step limit. With the previous code, it would work until length 80 (after creating a get_test_cases<80> function with similar patterns), but with the new impl, 63 is the max in those modes.

I'm not sure whether we should care? I've fixed the test by splitting up the test function.

Sample failure:
```
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp:129:17: error: static assertion expression is not an integral constant expression
# |   129 |   static_assert(test_to_string<64>());
# |       |                 ^~~~~~~~~~~~~~~~~~~~
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/build/generic-hardened-mode/include/c++/v1/string:1164:10: note: constexpr evaluation hit maximum step limit; possible infinite loop?
# |  1164 |         {return __is_long() ? __get_long_size() : __get_short_size();}
# |       |          ^
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/build/generic-hardened-mode/include/c++/v1/string:1205:50: note: in call to 'this->size()'
# |  1205 |     _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(__pos <= size(), "string index out of bounds");
# |       |                                                  ^
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp:38:20: note: in call to 's.operator[](42)'
# |    38 |             assert(s[b.size() - 1 - i] == one);
# |       |                    ^
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp:98:13: note: in call to 'check_equal(s, v, 48, 49)'
# |    98 |             check_equal(s, v, '0', '1');
# |       |             ^
# | /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-bd9c7138da7d-1/llvm-project/libcxx-ci/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp:129:17: note: in call to 'test_to_string()'
# |   129 |   static_assert(test_to_string<64>());
```

**Comment by philnik777** [comment]
@jyknight I don't think we have to worry much about that. The step count is quite arbitrary (maybe something like AST nodes?) and doesn't seem to have anything to do with the actual cost or compile time of the program. e.g. a `__builtin_memmove` of arbitrary size and a single assignment have the same step count AFAICT. You can avoid the test complexity by adding `// ADDITIONAL_COMPILE_FLAGS(has-fconstexpr-steps): -fconstexpr-steps=2000000` to the test to increase the step limit.

**Review by ldionne — APPROVED** [review]
Thanks a lot for working on this @jyknight!

**Comment by ldionne** [comment]
I'm merging now since it looks like everyone's happy with the PR and there are no outstanding comments.
